src/dotnet/ShellShimMaker/ShellShimMaker.cs

```diff
@@ -0,0 +1,98 @@
+using Microsoft.Extensions.EnvironmentAbstractions;
+
+namespace Microsoft.DotNet.ShellShimMaker
+{
+    public class ShellShimMaker
+    {
+        private readonly string _pathToPlaceShim;
+
+        public ShellShimMaker(string pathToPlaceShim)
+        {
+            _pathToPlaceShim =
+                pathToPlaceShim ?? throw new ArgumentNullException(nameof(pathToPlaceShim));
+        }
+
+        public void CreateShim(string packageExecutablePath, string shellCommandName)
+        {
+            var packageExecutable = new FilePath(packageExecutablePath);
+
+            var script = new StringBuilder();
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                script.AppendLine("@echo off");
+                script.AppendLine($"dotnet {packageExecutable.ToQuotedString()} %*");
+            }
+            else
+            {
+                script.AppendLine("#!/bin/sh");
+                script.AppendLine($"dotnet {packageExecutable.ToQuotedString()} \"$@\"");
+            }
+
+            var scriptPath = GetScriptPath(shellCommandName);
+            try
+            {
+                File.WriteAllText(scriptPath.Value, script.ToString());
+            }
+            catch (UnauthorizedAccessException e)
+            {
+                throw new GracefulException(
+                    string.Format("Install failed, try run the install command as an administrator, if you have the access. {0}",
+                        e.Message));
+            }
+
+            SetUserExecutionPermissionToShimFile(scriptPath);
+        }
+
+        public void EnsureCommandNameUniqueness(string shellCommandName)
+        {
+            if (File.Exists(Path.Combine(_pathToPlaceShim, shellCommandName)))
+            {
+                throw new GracefulException($"Failed to create tool {shellCommandName}, a command with the same name existed");
+            }
+        }
+
+        public void Remove(string shellCommandName)
+        {
+            File.Delete(GetScriptPath(shellCommandName).Value);
+        }
+
+        private FilePath GetScriptPath(string shellCommandName)
+        {
+            var scriptPath = Path.Combine(_pathToPlaceShim, shellCommandName);
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                scriptPath += ".cmd";
+            }
+
+            return new FilePath(scriptPath);
+        }
+
+        private static void SetUserExecutionPermissionToShimFile(FilePath scriptPath)
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) return;
+
+            var result = new CommandFactory()
+                .Create("chmod", new[] {"u+x", scriptPath.Value})
+                .CaptureStdOut()
+                .CaptureStdErr()
+                .Execute();
+
+
+            if (result.ExitCode != 0)
+            {
+                throw new GracefulException(
+                    "Failed to change permission" +
```

Comment:
Well, in general, we need to review these strings with @KathleenDollard, not a fan of this one either.

Reply:
Failed to change permission:
Error:
Output:

Reply:
Changed to Kathleen's version